scrapy_poet/commands.py

```diff
@@ -0,0 +1,115 @@
+import datetime
+from pathlib import Path
+from typing import Type
+
+import andi
+import scrapy
+from freezegun import freeze_time
+from scrapy import Request
+from scrapy.commands import ScrapyCommand
+from scrapy.crawler import Crawler
+from scrapy.exceptions import UsageError
+from scrapy.http import Response
+from scrapy.utils.misc import load_object
+from twisted.internet.defer import inlineCallbacks
+from web_poet import ItemPage
+from web_poet.testing import save_fixture
+
+from scrapy_poet import callback_for
+from scrapy_poet.downloadermiddlewares import DEFAULT_PROVIDERS, InjectionMiddleware
+from scrapy_poet.injection import Injector
+
+saved_dependencies = []
+saved_items = []
+
+
+class SavingInjector(Injector):
+    @inlineCallbacks
+    def build_instances_from_providers(
+        self, request: Request, response: Response, plan: andi.Plan
+    ):
+        instances = yield super().build_instances_from_providers(
+            request, response, plan
+        )
+        saved_dependencies.extend(instances.values())
+        return instances
+
+
+class SavingPipeline:
+    def process_item(self, item, spider):
+        saved_items.append(item)
+        return item
+
+
+class SavingInjectionMiddleware(InjectionMiddleware):
+    def __init__(self, crawler: Crawler) -> None:
+        super().__init__(crawler)
+        self.injector = SavingInjector(
+            crawler,
+            default_providers=DEFAULT_PROVIDERS,
+            overrides_registry=self.overrides_registry,
+        )
+
+
+def spider_for(injectable: Type[ItemPage]) -> Type[scrapy.Spider]:
+    class InjectableSpider(scrapy.Spider):
+        name = "injectable"
+
+        url = None
+        custom_settings = {
+            "SCRAPY_POET_PROVIDERS": DEFAULT_PROVIDERS,
+        }
+
+        def start_requests(self):
+            yield scrapy.Request(self.url, self.cb)
+
+        cb = callback_for(injectable)
+
+    return InjectableSpider
+
+
+def additional_settings() -> dict:
+    return {
+        "ITEM_PIPELINES": {
+            SavingPipeline: 100,
+        },
+        "DOWNLOADER_MIDDLEWARES": {
+            SavingInjectionMiddleware: 543,
+        },
```

Comment:
As a side effect of defining these, I take it we override project-configured values. This could be a good thing, but also a bad thing.

I cannot imagine an item pipeline that would be of use here, so I think disabling those could be good.

However, if a Scrapy project is using the scrapy-zyte-smartproxy plugin, I take it this effectively prevents that plugin from being used to download the URL, which I think could be a problem.

If so, we may need to use a more convoluted approach to define the downloader middlewares, i.e. add the middleware to the existing DOWNLOADER_MIDDLEWARES of self.settings below, instead of redefining DOWNLOADER_MIDDLEWARES.

Reply:
Makes sense.